File: src/main/java/main/commands/prefixcommands/audio/MoveCommand.java
package main.commands.prefixcommands.audio;

import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo;
import main.audiohandlers.RobertifyAudioManager;
import main.commands.prefixcommands.CommandContext;
import main.commands.prefixcommands.ICommand;
import main.utils.GeneralUtils;
import main.utils.RobertifyEmbedUtils;
import main.utils.json.dedicatedchannel.DedicatedChannelConfig;
import main.utils.json.logs.LogType;
import main.utils.json.logs.LogUtils;
import main.utils.locale.RobertifyLocaleMessage;
import net.dv8tion.jda.annotations.ForRemoval;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.internal.utils.tuple.Pair;

import javax.script.ScriptException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ConcurrentLinkedQueue;

@Deprecated @ForRemoval
public class MoveCommand implements ICommand {
    @Override
    public void handle(CommandContext ctx) throws ScriptException {
        final var musicManager = RobertifyAudioManager.getInstance().getMusicManager(ctx.getGuild());
        final var queue = musicManager.getScheduler().queue;
        final Message msg = ctx.getMessage();
        final List<String> args = ctx.getArgs();
        final var guild = ctx.getGuild();

        GeneralUtils.setCustomEmbed(ctx.getGuild(), "Queue");

        if (args.isEmpty()) {
            EmbedBuilder eb = RobertifyEmbedUtils.embedMessage(guild, "You must provide the ID of a song to move in the queue and the position to move it to.");
            msg.replyEmbeds(eb.build()).queue();
            return;
        } else if (args.size() < 2) {
            EmbedBuilder eb = RobertifyEmbedUtils.embedMessage(guild, "You must provide the position to move it the song to.");
            msg.replyEmbeds(eb.build()).queue();
            return;
        }

        if (!GeneralUtils.stringIsInt(args.get(0))) {
            EmbedBuilder eb = RobertifyEmbedUtils.embedMessage(guild, "You must provide a valid integer as the ID.");
            msg.replyEmbeds(eb.build()).queue();
            return;
        } else if (!GeneralUtils.stringIsInt(args.get(1))) {
            EmbedBuilder eb = RobertifyEmbedUtils.embedMessage(guild, "You must provide a valid integer as the position.");
            msg.replyEmbeds(eb.build()).queue();
            return;
        }

        final int id = Integer.parseInt(args.get(0));
        final int position = Integer.parseInt(args.get(1));

        msg.replyEmbeds(handleMove(ctx.getGuild(), ctx.getAuthor(), queue, id, position).build()).queue();

        GeneralUtils.setDefaultEmbed(ctx.getGuild());
    }

    public EmbedBuilder handleMove(Guild guild, User mover, ConcurrentLinkedQueue<AudioTrack> queue, int id, int position) {
        GeneralUtils.setCustomEmbed(guild, "Queue");

        if (queue.isEmpty())
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.NOTHING_IN_QUEUE);

        final List<AudioTrack> trackList = new ArrayList<>(queue);

        if (id <= 0 || id > trackList.size()) {
            EmbedBuilder eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.MoveMessages.INVALID_SONG_ID);
            return eb;
        } else if (position <= 0 || position > trackList.size()) {
            EmbedBuilder eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.MoveMessages.INVALID_POSITION_ID);
            return eb;
        }

        final List<AudioTrack> prevList = new ArrayList<>(queue);
        queue.clear();
        prevList.remove(trackList.get(id-1));
        prevList.add(position-1, trackList.get(id-1));
        if (!queue.addAll(prevList)) {
            queue.addAll(trackList);
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.MoveMessages.COULDNT_MOVE, Pair.of("{id}", String.valueOf(id)));
        }

        if (new DedicatedChannelConfig(guild).isChannelSet())
            new DedicatedChannelConfig(guild).updateMessage();

        AudioTrackInfo info = trackList.get(id - 1).getInfo();

        new LogUtils(guild).sendLog(LogType.TRACK_MOVE,
                RobertifyLocaleMessage.MoveMessages.MOVED_LOG,
                Pair.of("{user}", mover.getAsMention()),
                Pair.of("{title}", info.title),
                Pair.of("{author}", info.author),
                Pair.of("{position}", String.valueOf(position))
        );
        return RobertifyEmbedUtils.embedMessage(guild,
                RobertifyLocaleMessage.MoveMessages.MOVED,
                Pair.of("{title}", info.title),
                Pair.of("{author}", info.author),
                Pair.of("{position}", String.valueOf(position))
        );
    }

    @Override
    public String getName() {
        return "move";
    }

    @Override
    public String getHelp(String prefix) {
        return "Aliases: `"+GeneralUtils.listToString(getAliases())+"`\n" +
                "\nMove a specific track to a specific position in the queue\n" +
                "\nUsage: `"+ prefix+"move <id> <position>`";
    }

    @Override
    public List<String> getAliases() {
        return List.of("m");
    }
}

File: src/main/kotlin/main/commands/slashcommands/audio/MoveCommand.kt
package main.commands.slashcommands.audio

import main.audiohandlers.QueueHandler
import main.audiohandlers.RobertifyAudioManager
import main.audiohandlers.utils.author
import main.audiohandlers.utils.title
import main.commands.slashcommands.SlashCommandManager.getRequiredOption
import main.utils.RobertifyEmbedUtils
import main.utils.RobertifyEmbedUtils.Companion.replyEmbed
import main.utils.RobertifyEmbedUtils.Companion.sendEmbed
import main.utils.component.interactions.slashcommand.AbstractSlashCommand
import main.utils.component.interactions.slashcommand.models.SlashCommand
import main.utils.component.interactions.slashcommand.models.CommandOption
import main.utils.json.logs.LogType
import main.utils.json.logs.LogUtilsKt
import main.utils.json.requestchannel.RequestChannelConfig
import main.utils.locale.messages.GeneralMessages
import main.utils.locale.messages.MoveMessages
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.commands.OptionType

class MoveCommand : AbstractSlashCommand(
    SlashCommand(
        name = "move",
        description = "Rearrange the position of tracks in the queue.",
        options = listOf(
            CommandOption(
                type = OptionType.INTEGER,
                name = "id",
                description = "The ID of the track in the queue to move."
            ),
            CommandOption(
                type = OptionType.INTEGER,
                name = "position",
                description = "The position to move the track to."
            )
        )
    )
) {

    override suspend fun handle(event: SlashCommandInteractionEvent) {
        val guild = event.guild!!
        val memberVoiceState = event.member!!.voiceState!!
        val selfVoiceState = guild.selfMember.voiceState!!
        val acChecks = audioChannelChecks(memberVoiceState, selfVoiceState)
        if (acChecks != null) {
            event.replyEmbed { acChecks }.setEphemeral(true).queue()
            return
        }

        event.deferReply().queue()
        val queueHandler = RobertifyAudioManager[guild]
            .scheduler
            .queueHandler
        val id = event.getRequiredOption("id").asInt
        val pos = event.getRequiredOption("position").asInt

        event.hook.sendEmbed {
            handleMove(
                guild = guild,
                mover = event.user,
                queueHandler = queueHandler,
                id = id,
                position = pos
            )
        }.queue()
    }

    private fun handleMove(
        guild: Guild,
        mover: User,
        queueHandler: QueueHandler,
        id: Int,
        position: Int
    ): MessageEmbed {
        if (queueHandler.isEmpty)
            return RobertifyEmbedUtils.embedMessage(guild, GeneralMessages.NOTHING_IN_QUEUE)
                .build()

        val trackList = queueHandler.contents.toMutableList()

        if (id <= 0 || id > trackList.size)
            return RobertifyEmbedUtils.embedMessage(guild, MoveMessages.INVALID_SONG_ID)
                .build()
        else if (position <= 0 || position > trackList.size)
            return RobertifyEmbedUtils.embedMessage(guild, MoveMessages.INVALID_POSITION_ID)
                .build()

        val prevList = trackList.toMutableList()
        queueHandler.clear()
        prevList.remove(trackList[id - 1])
        prevList.add(position - 1, trackList[id - 1])

        if (!queueHandler.addAll(prevList)) {
            queueHandler.addAll(trackList)
            return RobertifyEmbedUtils.embedMessage(
                guild,
                MoveMessages.COULDNT_MOVE,
                Pair("{id}", id.toString())
            ).build()
        }

        RequestChannelConfig(guild).updateMessage()

        val movedTrack = trackList[id - 1]
        LogUtilsKt(guild).sendLog(
            LogType.TRACK_MOVE,
            MoveMessages.MOVED_LOG,
            Pair("{user}", mover.asMention),
            Pair("{title}", movedTrack.title),
            Pair("{author}", movedTrack.author),
            Pair("{position}", position.toString())
        )

        return RobertifyEmbedUtils.embedMessage(
            guild,
            MoveMessages.MOVED,
            Pair("{title}", movedTrack.title),
            Pair("{author}", movedTrack.author),
            Pair("{position}", position.toString())
        ).build()
    }

    override val help: String
        get() = """
                Move a specific track to a specific position in the queue

                Usage: `/move <id> <position>`"""

}